woocommerce/woocommerce
# Introduce InMemoryCacheEngine for use by Order Cache.

### Changes proposed in this Pull Request:

Closes #45550.

This PR introduces a RequestLevelCacheEngine implementation of `Automattic\WooCommerce\Caching\CacheEngine` and replaces the normal WPCacheEngine used by OrderCache with this implementation.

The RequestLevelCacheEngine works as an in-memory cache only and does not run through the `wp_cache_*` functions.

Because of the serialization method of Orders, the first time an Order is loaded from cache, the data is retrieved from the datastore.  This results in worse performance when querying multiple orders than if no persistent cache is available at all because the data for each order is queried one at a time as opposed as to all at once.

The use of a request only cache provides the performance benefit of storing the order objects in memory so they don't need to be required in the same request without persisting them in cache causing them to go through serialization - defeating the benefit of the cache.

This is an initial step for implementing #46023 but has shown some slight performance improvements on its own.  Testing against a site running WooCommerce Subscriptions and Memcached with 15K+ orders has shown to improve decrease request times by 0% - 5%. Again, this is small on its own, but is necessary for adding order data caching (#46023) as an optional feature which has shown reduced request times by 30% - 50%.




### How to test the changes in this Pull Request:

1. Enable HPOS with the HPOS tables as authoritative to enable Order Caching.
2. In the wp-admin, load the WooCommerce -> Orders page, verify that orders appear correctly as listed.
3. Open an individual order and make a modification such as editing meta, an address, status, etc. and save.
4. Validate that the data saved properly and the reloaded order appears correctly.
5. Place an order through the front-end of the store.
6. Verify that the checkout process completed correctly and data was properly updated when reviewing the order in the admin.

### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Introduce InMemoryCacheEngine for use by Order Cache.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
